### PR TITLE
dsl2 port

### PR DESCRIPTION
**Pull request by phue**
Continuation of https://github.com/nf-core/methylseq/pull/182 (now from a dedicated branch)

:warning: Breaking change:

The pipeline now requires a sample sheet to be passed to the pipeline with `--input`:

| sample | fastq_1 | fastq_2 | genome |
|--- | --- | --- | --- |

See an example [here](https://github.com/nf-core/test-datasets/blob/methylseq/samplesheet.csv)

The `genome` column is optional, `--genome` continues to work as well 

Known issues:

- [x] ~workflow summary shows NAs~ -> (fixed by @ewels in https://github.com/nf-core/methylseq/pull/206)
- [x]  ~`MultiQC` report is missing some data~ -> (fixed in 0fd1614)
- [x] ~`-profile test_full` needs a `samplesheet.csv`~ -> (fixed in https://github.com/nf-core/test-datasets/pull/232)

### CODE REVIEW

**Comment by ewels** [comment]
> _workflow summary shows NAs_

Maybe you mean as described in https://github.com/nf-core/tools/issues/971 ? In which case, upstream fix is required.

**Comment by rbpisupati** [comment]
Thanks for this update @phue ! It looks great 🎉 I did try this and found some errors/fixes

1. Gives an error when few samples are single-end and others paired-end. Would be great if it can detect presence of second pair.
2. Also, the output bam files in `deduplicated` folder inside `bismark` has multiple files with names `sample_id_1_val_1_bismark_bt2_pe.deduplicated.bam` and `sample_id.deduplicated.sorted.bam` . It is bit confusing and would be great if you can only put the final bam files in. 
3. There seems to be some error in $multicore variable in `modules/local/software/bismark/align/main.nf` .. it seems to resolve when I remove that. 



Cheers, 
R

**Comment by phue** [comment]
Hi @rbpisupati,

thanks alot for testing this.

>     1. Gives an error when few samples are single-end and others paired-end. Would be great if it can detect presence of second pair.

Would you be able to share some more information on this? Where does the error occur exactly?

>     2. Also, the output bam files in `deduplicated` folder inside `bismark` has multiple files with names `sample_id_1_val_1_bismark_bt2_pe.deduplicated.bam` and `sample_id.deduplicated.sorted.bam` . It is bit confusing and would be great if you can only put the final bam files in.

Fixed in 5ad41fe (it copied both the sorted and the unsorted bam into the results directory. It should be enough to keep the former)

>     3. There seems to be some error in $multicore variable in `modules/local/software/bismark/align/main.nf` .. it seems to resolve when I remove that.

Fixed in df97fa1

**Comment by rbpisupati** [comment]
If I remember correctly, it is trying to do paired-end mapping with filename `null` since file path isnt present in the `samplesheet.csv`.

**Comment by phue** [comment]
> If I remember correctly, it is trying to do paired-end mapping with filename `null` since file path isnt present in the `samplesheet.csv`.

I tried to reproduce reproduce it with this [samplesheet](https://gist.github.com/phue/7a422d73621773800d0372c2cae4e8fd) but it works as expected. How does your samplesheet look like?

**Comment by ewels** [comment]
Ack, I tried to bring this DSL2 branch back up to date with `dev`, but now tests are failing and I have no idea why 😰 

@phue / anyone else have some time to look at this?

File: .nf-core-lint.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# FIXME - remove after DSL2 template is released in tools
+actions_awsfulltest: False
```
Comment (ewels) [inline .nf-core-lint.yaml:2]:
```suggestion
```

**Comment by ewels** [comment]
ok I started trying to bring this up to date in https://github.com/nf-core/methylseq/pull/222

I did it as a PR to the `dsl2` branch so that I / others can see the changes I've made to the original work from @phue. Hopefully once I get that working we can then move through the remaining PRs to a release relatively quickly.

**Comment by njspix** [comment]
I think (hope) this is ready to merge into dev. I did make one substantial plumbing change - for the bismark workflow, I elected to sort the bam immediately after mapping. This mirrors the bwa-meth workflow and seems to make everything a bit more consistent. Also, this ensures that preseq receives *sorted* bam

File: modules/local/functions.nf [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (Emiller88) [inline modules/local/functions.nf:1]: Is this file necessary?

File: workflows/methylseq.nf [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+//
+// SUBWORKFLOW: Consisting of a mix of local and nf-core/modules
+//
+include { INPUT_CHECK } from '../subworkflows/local/input_check'
+
+if( params.aligner =~ /bismark/ ){
+    include { BISMARK as ALIGNER } from '../subworkflows/local/bismark'
+} else if ( params.aligner == 'bwameth' ){
+    include { BWAMETH as ALIGNER } from '../subworkflows/local/bwameth'
+}
```
Comment (Emiller88) [inline workflows/methylseq.nf:82]: This is really clever, I haven't seen it done in any nf-core workflows before, so I'm not sure if it'll be transparent enough to anyone reading the workflow.

**Review by Emiller88 — APPROVED** [review]
Looks good to me, just one thing to double-check if the `functions.nf` file is doing anything, and your aligner import was super clever, I just don't know if people will miss it.

**Comment by phue** [comment]
Hi @njspix,

thanks a bunch for picking this up and pushing it forward, great job!

Though I have to say I'm not particularly happy with the fact that the "one-sample-one-genome"-functionality that I put in to address https://github.com/nf-core/methylseq/issues/181 was lost along the way, I think we had agreed to disagree on that one (x-ref https://github.com/nf-core/modules/pull/129#issuecomment-774755102 https://github.com/nf-core/methylseq/pull/222#issuecomment-961871739).  

I have a more serious concern however, about the changes made to the `bismark` workflow:
If I remember correctly, the bismark tooling requires the bam file to be name-sorted, i.e `bismark-methylation-extractor` requires paired-end reads to consecutively follow each other in the input bam.

Did you test this on full-sized paired-end data (ideally with and without deduplication) ?

**Comment by njspix** [comment]
@phue Thanks for reviewing! I am sorry about the per-sample genome functionality, I figured it was probably best to at least start with just basic functionality for a first release. I would be happy to help add in that feature later (I had some ideas about how to structure the channels so we wouldn't have to patch the aligner modules). 

Thanks for raising the concern about bismark name sorting. I'll check and make sure.

**Comment by ewels** [comment]
Yup, Bismark needs BAMs to be sorted by name: https://github.com/FelixKrueger/Bismark/tree/master/Docs#iii-bismark-deduplication-step

I also am partly to blame for dropping the per-sample functionality. I was pushing to not have local copies and made some headway with building new tooling to support patched versions of central modules. But life happened and I never finished it (maybe something else that I can delegate to @mirpedrol at some point!).

Let's keep it as a high priority to add in ASAP via either my or @njspix method. But for now the top priority must be DSL2 conversion and release I think 👍🏻

**Comment by phue** [comment]
Don't get me wrong I'm not blaming anyone here, I think what happened is that it got decided by popular vote a while ago. So fine with me, although I would have liked to see that functionality.

Maybe it will come up in other pipelines as well at some point and the necessary changes will make it to `nf-core modules` so it can be easily added back here (without the need for local modules to implement it)

**Comment by Emiller88** [comment]
If we're making a wishlist, I'd love to handle different kits in the same workflow run!  :grin: 
I think the samplesheet opens up for this in the future.

On the topic of #181, maybe we could add the additional fasta functionality in rnaseq. I `cat`'d them together manually ahead of time to make a reference.

**Comment by ewels** [comment]
Let's do it! :shipit: 

lightspeed

**Review by ewels — APPROVED** [review]
